rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,662 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use arrow::array::ArrayData;
+use arrow::datatypes::{ArrowPrimitiveType, UInt16Type, UInt32Type, UInt64Type, UInt8Type};
+use arrow_array::{cast::AsArray, Array, ArrayRef, PrimitiveArray};
+use arrow_schema::DataType;
+use bytes::{Bytes, BytesMut};
+use futures::future::{BoxFuture, FutureExt};
+use log::trace;
+use num_traits::{AsPrimitive, PrimInt};
+use snafu::{location, Location};
+
+use lance_arrow::DataTypeExt;
+use lance_core::{Error, Result};
+
+use crate::encoder::EncodedBufferMeta;
+use crate::{
+    decoder::{PageScheduler, PrimitivePageDecoder},
+    encoder::{BufferEncoder, EncodedBuffer},
+};
+
+// Compute the number of bits to use for each item, if this array can be encoded using
+// bitpacking encoding. Returns `None` if the type or array data is not supported.
+pub fn num_compressed_bits(arr: ArrayRef) -> Option<u64> {
+    match arr.data_type() {
+        DataType::UInt8 => num_bits_for_type::<UInt8Type>(arr.as_primitive()),
+        DataType::UInt16 => num_bits_for_type::<UInt16Type>(arr.as_primitive()),
+        DataType::UInt32 => num_bits_for_type::<UInt32Type>(arr.as_primitive()),
+        DataType::UInt64 => num_bits_for_type::<UInt64Type>(arr.as_primitive()),
+        // TODO -- eventually we could support signed types as well
+        _ => None,
+    }
+}
+
+// Compute the number of bits to to use for bitpacking generically.
+// Returns `None` if the array is empty or only contains null values.
+fn num_bits_for_type<T>(arr: &PrimitiveArray<T>) -> Option<u64>
+where
+    T: ArrowPrimitiveType,
+    T::Native: PrimInt + AsPrimitive<u64>,
+{
+    let max = arrow::compute::bit_or(arr);
```

Comment:
Minor nit:  We can probably assume arrays are not empty.  Also, I think we have an "all values null" filter in `BasicEncoder`.  So by this point it's probably safe to unwrap.

Reply:
Yeah that works. Made this change